drivers/sensor/bosch/bmm350/bmm350_trigger.c

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2024 Bosch Sensortec GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/*
+ * Bus-specific functionality for BMM350s accessed via I2C.
+ */
+#include <zephyr/kernel.h>
+#include <zephyr/pm/device.h>
+#include <zephyr/logging/log.h>
+
+#include "bmm350.h"
+
+LOG_MODULE_DECLARE(BMM350, CONFIG_SENSOR_LOG_LEVEL);
+
+static void bmm350_handle_interrupts(const void *arg)
+{
+	const struct device *dev = (const struct device *)arg;
+	struct bmm350_data *data = dev->data;
+
+	if (data->drdy_handler) {
+		data->drdy_handler(dev, data->drdy_trigger);
+	}
+}
+
+#ifdef CONFIG_BMM350_TRIGGER_OWN_THREAD
+static K_THREAD_STACK_DEFINE(bmm350_thread_stack,
+			     CONFIG_BMM350_THREAD_STACK_SIZE);
+static struct k_thread bmm350_thread;
+
+static void bmm350_thread_main(void *arg1, void *unused1, void *unused2)
+{
+	ARG_UNUSED(unused1);
+	ARG_UNUSED(unused2);
+	const struct device *dev = (const struct device *)arg1;
+	struct bmm350_data *data = dev->data;
+
+	while (1) {
+		k_sem_take(&data->sem, K_FOREVER);
+		bmm350_handle_interrupts(dev);
+	}
+}
+#endif
+
+#ifdef CONFIG_BMM350_TRIGGER_GLOBAL_THREAD
+static void bmm350_work_handler(struct k_work *work)
+{
+	struct bmm350_data *data = CONTAINER_OF(work,
+						struct bmm350_data,
+						work);
+
+	bmm350_handle_interrupts(data->dev);
+}
+#endif
+
+static void bmm350_gpio_callback(const struct device *port,
+				 struct gpio_callback *cb,
+				 uint32_t pin)
+{
+	struct bmm350_data *data = CONTAINER_OF(cb,
+						struct bmm350_data,
+						gpio_cb);
+	ARG_UNUSED(port);
+	ARG_UNUSED(pin);
+
+#if defined(CONFIG_BMM350_TRIGGER_OWN_THREAD)
+	k_sem_give(&data->sem);
+#elif defined(CONFIG_BMM350_TRIGGER_GLOBAL_THREAD)
+	k_work_submit(&data->work);
+#elif defined(CONFIG_BMM350_TRIGGER_DIRECT)
+	bmm350_handle_interrupts(data->dev);
```

Comment:
You absolutely cannot do this. This will call the handler on the interrupt context.

Reply:
I removed the TRIGGER_DIRECT stuff. Other driver still have this though.